### ISSUE TITLE
async-coap-uri: no_std implementation is broken

### ISSUE THREAD

**Issue by Luro02** [google/rust-async-coap#26]
The crate provides the following features:
```toml
[features]
default = ["std"]
std = ["alloc"]
alloc = []
```

by default `std` is enabled, which works as expected.

To test the features I simply ran the tests of `async-coap-uri` with the below commands.

To disable all default features one could try the following command (this would be true `no_std` without any allocator):

```
cargo test --no-default-features
```

which fails (with lots of errors).

Enabling only the `alloc`-feature also fails

```
cargo test --no-default-features --features alloc
```

**Comment by darconeous** [comment]
Yep, this is a bug. Architecturally, rust-async-coap is designed to eventually allow for use in `no_std` mode without even `alloc`. However, getting things to work without `alloc` is going to require [generic associated types](https://github.com/rust-lang/rfcs/blob/master/text/1598-generic_associated_types.md) to land, so I didn't bother too much with it other than to make sure that architecturally it wasn't precluded from being introduced later.

However, there is no good reason for not supporting `no_std` with `alloc` right now. Getting that implemented would be a good step toward reaching that eventual goal once GAT lands.
